daemon/applications.c

```diff
@@ -0,0 +1,465 @@
+static appinfo_t *applications_add_appinfo   (applications_t *self, const char *appname);
+static bool       applications_remove_appinfo(applications_t *self, const char *appname);
+
+/* ========================================================================= *
+ * APPLICATIONS
+ * ========================================================================= */
+
+struct applications_t
+{
+    bool                   aps_initialized;
+    control_t             *aps_control;
+    stringset_t           *aps_available;
+    guint                  aps_rescan_id;
+    GFileMonitor          *aps_monitor_obj;
+    GHashTable            *aps_appinfo_lut;
+};
+
+static void
+applications_ctor(applications_t *self, control_t *control)
+{
+    log_info("applications() create");
+
+    self->aps_initialized = false;
+    self->aps_control     = control;
+    self->aps_available   = stringset_create();
+    self->aps_rescan_id   = 0;
+    self->aps_monitor_obj = NULL;
+    self->aps_appinfo_lut = g_hash_table_new_full(g_str_hash,
+                                                  g_str_equal,
+                                                  g_free,
+                                                  appinfo_delete_cb);
+
+    /* Fetch initial state */
+    applications_start_monitor(self);
+    applications_scan_now(self);
+
+    self->aps_initialized = true;
+}
+
+static  void
+applications_dtor(applications_t *self)
+{
+    log_info("applications() delete");
+
+    self->aps_initialized = false;
+
+    applications_stop_monitor(self);
+    applications_cancel_rescan(self);
+
+    if( self->aps_appinfo_lut ) {
+        g_hash_table_unref(self->aps_appinfo_lut),
+            self->aps_appinfo_lut = NULL;
+    }
+
+    stringset_delete_at(&self->aps_available);
+}
+
+applications_t *
+applications_create(control_t *control)
+{
+    applications_t *self = g_malloc0(sizeof *self);
+    applications_ctor(self, control);
+    return self;
+}
+
+void
+applications_delete(applications_t *self)
+{
+    if( self ) {
+        applications_dtor(self);
+        g_free(self);
+    }
+}
+
+void
+applications_delete_at(applications_t **pself)
+{
+    applications_delete(*pself), *pself = NULL;
+}
+
+void
+applications_delete_cb(void *self)
+{
+    applications_delete(self);
+}
+
+/* ------------------------------------------------------------------------- *
+ * APPLICATIONS_ATTRIBUTES
+ * ------------------------------------------------------------------------- */
+
+control_t *
+applications_control(const applications_t *self)
+{
+    return self->aps_control;
+}
+
+const stringset_t *
+applications_available(applications_t *self)
+{
+    if( applications_cancel_rescan(self) )
+        applications_scan_now(self);
+    return self->aps_available;
+}
+
+appinfo_t *
+applications_appinfo(applications_t *self, const char *appname)
+{
+    appinfo_t *appinfo = applications_appinfo(self, appname);
```

Comment:
Very suspicious looking recursion here.

Reply:
Well, just bumped into it after running the binary instead of just checking that it compiles ;-)
Fixed.